core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -600,10 +600,30 @@ private[spark] object Utils extends Logging {
     if (lowerSrc.endsWith(".jar")) {
       RunJar.unJar(source, dest, RunJar.MATCH_ANY)
     } else if (lowerSrc.endsWith(".zip")) {
+      // TODO(SPARK-37677): should keep file permissions. Java implementation doesn't.
       FileUtil.unZip(source, dest)
-    } else if (
-      lowerSrc.endsWith(".tar.gz") || lowerSrc.endsWith(".tgz") || lowerSrc.endsWith(".tar")) {
+    } else if (lowerSrc.endsWith(".tar.gz") || lowerSrc.endsWith(".tgz")) {
       FileUtil.unTar(source, dest)
+    } else if (lowerSrc.endsWith(".tar")) {
+      // The codes below was copied from `FileUtil.unTar` but uses Java-based implementation
+      // to work around a security issue, see also SPARK-38631.
+      // TODO(SPARK-38632): should keep file permissions. Java implementation doesn't.
+      if (!dest.mkdirs && !dest.isDirectory) {
```

Comment:
Only nit is then - maybe make this a private method for symmetry with the existing FileUtil.untar?

Reply:
Sounds good!